tests/README.md

```diff
@@ -0,0 +1,1510 @@
+Subtypes should be substitutable for their base types. From a clients perspective overriding methods should not break
+functionality.
+<a name="interface-segregation-principle-isp"></a>
+##### Interface Segregation Principle (ISP)
+Clients should not be forced to depend on methods they don't use. Split a larger interface into
+a number of smaller interfaces.
+<a name="dependency-inversion-principal-dip"></a>
+##### Dependency Inversion Principal (DIP)
+High-level modules should not depend on low-level modules; both should depend on abstractions.
+Abstractions should not depend on details; details should depend on abstractions.
+
+<a name="a-solid-way-to-use-the-dic"></a>
+#### A SOLID way to use the DIC
+A good way to improve the testability of new and old classes are the inversion of the dependencies.
+For example the classic way to use another class is:  
+
+```php
+<?php
+class Car {
+
+	private $breaks;
+
+	public function __construct() {
+		$this->breaks = new StandardBreaks();
+	}
+
+	public function stop() { /* use the breaks ... */}
+}
+```
+To verify that a car is able to stop, the breaks have to be replaced to verify the behaviour because the
+stop function has no return value. However there is no way to replace the hardwired dependency to the breaks
+except to load a class with the same name before actual class is loaded which could be considered a hack.
+
+In order to increase the testability the hardwired dependency has to be inverted (DIP).
+- First a Breaks interface is created for the StandardBreaks.
+- The Breaks will be passed to the Car at construction time. (The factories from Audi etc. do the same.)
+
+After these changes the car don't care about the actual implementation because it only depends on
+the breaks abstraction. This changes allows the car developers to finally test the class because the breaks can be
+exchanged at test time without hassle.
+
+```php
+<?php
+class Car {
+
+	private $breaks;
+
+	public function __construct(Breaks $breaks) {
+		$this->breaks = $breaks;
+	}
+
+	public function stop() { /* use the breaks ... */}
+}
+```
+
+The corresponding test class would look like this:
+
+```php
+<?php
+use \PHPUnit\Framework\TestCase;
+use \Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
+
+class CarTest extends TestCase {
+
+	use MockeryPHPUnitIntegration;
+
+	private $subject, $breaks;
+	/**
+	 * @inheritDoc
+	 */
+	protected function setUp()
+	{
+		parent::setUp();
+
+		$this->breaks = Mockery::mock(Breaks::class);
+		$this->subject = new Car($this->breaks);
+	}
+
+	/**
+	 * @test
+	 */
+	public function testStopCarWhileDriving() { /* use the breaks mock to verify behaviour ... */}
+}
+```
+
+This principle could improve the current situation in ILIAS a lot because many classes are hardwired and therefore
+not testable at all. In almost any part in the old code structures in ILIAS serves the DIC as nothing more as a service locator
+which is technically the same than the globals used before.
```

Comment:
This is correct. We currently are on a long way to replace globals by dependency injection, service locator is just an intermediate step there.